lib/iex/lib/iex/evaluator.ex

```diff
@@ -279,18 +309,21 @@ defmodule IEx.Evaluator do
     end
   end
 
-  defp do_eval({:ok, forms, buffer}, iex_state, state) do
+  defp do_eval({:ok, forms, buffer}, iex_state, state) when is_binary(buffer),
+    do: do_eval({:ok, forms, {buffer, :other}}, iex_state, state)
+
+  defp do_eval({:ok, forms, {buffer, last_op}}, iex_state, state) do
```

Comment:
Sorry, to be clear, the evaluator should not know anything about `last_op`. It is a parser specific behaviour that we shouldn't leak. This should be an "opaque" term to the evaluator: 

```suggestion
  defp do_eval({:ok, forms, parser_state}, iex_state, state) do
```

My suggestion is to rename the `:buffer` key to `:parser_state`.

Reply:
Ah. Got it. ok.

Reply:
@josevalim could you please provide guidance on where would I store the `last_op` between different `parse/3` invocations, if not in evaluator’s state? This:
```elixir
iex|1> x = 42
iex|2> |> IO.puts()
```
results in two invocations of a pure `parse/3` function and the second invocation should take the value of `last_op` set during the first one. Am I missing something?

Reply:
I think I finally get it.

> rename the `:buffer` key to `:parser_state`

`:buffer` key belongs to `IEx.State`. I was under the impression you didn’t want to amend `IEx.State` and keep things private to the evaluator. But we can move it there, of course.

Reply:
I think I finally get it.

> rename the `:buffer` key to `:parser_state`

`:buffer` key belongs to `IEx.State`. I was under the impression you didn’t want to amend `IEx.State` and keep things private to the evaluator. But we can move it there, of course.

Reply:
I couldn't answer your question without trying the code, so I just pushed a change that makes the parser state opaque to the server: https://github.com/elixir-lang/elixir/commit/abc23d00c4a9bbafcf08237ed64f93e175196bfa

Please rebase and you should be able to address this by only changing the parse function. :)